File: cpp_core/tests/test_factors.cpp
#include "volarb/factors.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include <cmath>
#include <vector>

using Catch::Approx;
using volarb::correlation_of;
using volarb::decompose_surface_factors;
using volarb::effective_sample_size;
using volarb::FactorDecomposition;
using volarb::inner_product;
using volarb::InvalidFactorInputsError;
using volarb::lag_one_autocorrelation;
using volarb::neutralise_against_loadings;
using volarb::orthonormal_basis;
using volarb::raw_factor_vectors;
using volarb::score_residual;
using volarb::SurfacePoint;

namespace {

const std::vector<double> tenors = {0.0833, 0.25, 0.5, 1.0, 2.0};
const std::vector<double> strikes = {-0.4, -0.2, -0.1, 0.0, 0.1, 0.2, 0.4};

std::vector<SurfacePoint> full_grid() {
    std::vector<SurfacePoint> grid;
    for (const double tenor : tenors) {
        for (const double strike : strikes) {
            grid.push_back(SurfacePoint{strike, tenor});
        }
    }
    return grid;
}

std::vector<std::vector<double>> factor_driven_surfaces(int count) {
    std::vector<std::vector<double>> observations;
    for (int index = 0; index < count; ++index) {
        const double phase = static_cast<double>(index);
        const double level = 0.04 * std::exp(0.2 * std::sin(phase * 0.31));
        const double slope = 1.0 + 0.1 * std::cos(phase * 0.17);
        const double skew = -0.35 + 0.08 * std::sin(phase * 0.23);
        const double curvature = 0.9 + 0.15 * std::cos(phase * 0.11);
        std::vector<double> surface;
        for (const double tenor : tenors) {
            for (const double strike : strikes) {
                surface.push_back(level * std::pow(tenor, slope) *
                                  std::exp(skew * strike + curvature * strike * strike));
            }
        }
        observations.push_back(surface);
    }
    return observations;
}

}

TEST_CASE("the named basis is orthonormal on the grid", "[factors]") {
    const std::vector<std::vector<double>> basis = orthonormal_basis(raw_factor_vectors(full_grid()));
    REQUIRE(basis.size() == 4);
    for (std::size_t row = 0; row < basis.size(); ++row) {
        for (std::size_t column = 0; column < basis.size(); ++column) {
            const double expected = row == column ? 1.0 : 0.0;
            REQUIRE(inner_product(basis[row], basis[column]) == Approx(expected).margin(1e-12));
        }
    }
}

TEST_CASE("a surface built only from the factors leaves no residual", "[factors]") {
    const FactorDecomposition decomposition =
        decompose_surface_factors(full_grid(), factor_driven_surfaces(120));
    REQUIRE(decomposition.identified_factor_count == 4);
    REQUIRE(decomposition.variance_explained > 0.999999);
    for (const std::vector<double>& residual : decomposition.residuals) {
        for (const double value : residual) {
            REQUIRE(std::abs(value) < 1e-12);
        }
    }
}

TEST_CASE("residuals are orthogonal to the basis at every observation", "[factors]") {
    const std::vector<SurfacePoint> grid = full_grid();
    const FactorDecomposition decomposition =
        decompose_surface_factors(grid, factor_driven_surfaces(60));
    const std::vector<std::vector<double>> basis = orthonormal_basis(raw_factor_vectors(grid));
    for (const std::vector<double>& residual : decomposition.residuals) {
        for (const std::vector<double>& vector : basis) {
            REQUIRE(inner_product(residual, vector) == Approx(0.0).margin(1e-10));
        }
    }
}

TEST_CASE("a single expiry grid cannot identify a term slope", "[factors]") {
    std::vector<SurfacePoint> grid;
    for (const double strike : strikes) {
        grid.push_back(SurfacePoint{strike, 0.25});
    }
    std::vector<std::vector<double>> observations;
    for (int index = 0; index < 20; ++index) {
        std::vector<double> surface;
        for (const double strike : strikes) {
            surface.push_back(0.04 * std::exp(-0.3 * strike + 0.02 * static_cast<double>(index)));
        }
        observations.push_back(surface);
    }
    const FactorDecomposition decomposition = decompose_surface_factors(grid, observations);
    REQUIRE(decomposition.identified_factor_count == 3);
    for (const volarb::FactorLoadings& entry : decomposition.loadings) {
        REQUIRE(entry.curvature == Approx(0.0));
    }
}

TEST_CASE("time domain neutralisation removes what the cross section leaves", "[factors]") {
    const std::vector<SurfacePoint> grid = full_grid();
    std::vector<std::vector<double>> observations = factor_driven_surfaces(200);
    for (std::size_t index = 0; index < observations.size(); ++index) {
        observations[index][3] *= 1.0 + 0.02 * std::sin(static_cast<double>(index) * 0.31);
    }
    const FactorDecomposition decomposition = decompose_surface_factors(grid, observations);

    std::vector<double> series;
    for (const std::vector<double>& residual : decomposition.residuals) {
        series.push_back(residual[3]);
    }
    const volarb::NeutralisedResidual neutralised =
        neutralise_against_loadings(series, decomposition.loadings,
                                    decomposition.identified_factor_count);
    REQUIRE(neutralised.worst_factor_correlation < 1e-12);
    REQUIRE(neutralised.worst_factor_correlation < neutralised.worst_factor_correlation_before);
}

TEST_CASE("autocorrelation inflates the standard error of a z score", "[factors]") {
    for (const double rho : {0.0, 0.5, 0.8, 0.9}) {
        std::vector<double> series;
        double value = 0.0;
        double state = 1.0;
        for (int index = 0; index < 400; ++index) {
            state = std::fmod(state * 48271.0, 2147483647.0);
            const double shock = (state / 2147483647.0 - 0.5) * std::sqrt(12.0);
            value = rho * value + std::sqrt(1.0 - rho * rho) * shock;
            series.push_back(value);
        }
        const volarb::ResidualScore score = score_residual(series);
        const bool persistent = score.lag_one_autocorrelation > 0.0;
        REQUIRE((score.naive_overstatement > 1.0) == persistent);
        REQUIRE((score.effective_sample_size < static_cast<double>(series.size())) == persistent);
        if (persistent) {
            REQUIRE(std::abs(score.adjusted_z_score) < std::abs(score.naive_z_score));
        }
        if (rho > 0.5) {
            REQUIRE(score.lag_one_autocorrelation > 0.0);
            REQUIRE(score.naive_overstatement > 1.0);
        }
    }
}

TEST_CASE("the effective sample size follows the autocorrelation", "[factors]") {
    REQUIRE(effective_sample_size(100, 0.0) == Approx(100.0));
    REQUIRE(effective_sample_size(100, 0.9) == Approx(100.0 * 0.1 / 1.9));
    const std::vector<double> constant(50, 3.0);
    REQUIRE(lag_one_autocorrelation(constant) == Approx(0.0));
}

TEST_CASE("malformed factor inputs are rejected", "[factors]") {
    const std::vector<SurfacePoint> grid = full_grid();
    REQUIRE_THROWS_AS(decompose_surface_factors({SurfacePoint{0.0, 1.0}}, {{1.0}}),
                      InvalidFactorInputsError);
    REQUIRE_THROWS_AS(decompose_surface_factors(grid, {}), InvalidFactorInputsError);
    REQUIRE_THROWS_AS(decompose_surface_factors(grid, {std::vector<double>(grid.size(), 0.04),
                                                      std::vector<double>(3, 0.04)}),
                      InvalidFactorInputsError);
    REQUIRE_THROWS_AS(decompose_surface_factors(grid, {std::vector<double>(grid.size(), -0.04),
                                                      std::vector<double>(grid.size(), 0.04)}),
                      InvalidFactorInputsError);
    REQUIRE_THROWS_AS(score_residual({1.0}), InvalidFactorInputsError);
}

TEST_CASE("a residual uncorrelated with the loadings is unchanged in correlation", "[factors]") {
    const FactorDecomposition decomposition =
        decompose_surface_factors(full_grid(), factor_driven_surfaces(80));
    std::vector<double> series(decomposition.loadings.size(), 0.0);
    for (std::size_t index = 0; index < series.size(); ++index) {
        series[index] = std::sin(static_cast<double>(index) * 1.7);
    }
    const volarb::NeutralisedResidual neutralised = neutralise_against_loadings(
        series, decomposition.loadings, decomposition.identified_factor_count);
    REQUIRE(neutralised.worst_factor_correlation < 1e-12);
    REQUIRE(correlation_of(neutralised.values, series) > 0.5);
}

File: python_pure/tests/test_factors.py
from __future__ import annotations

import math

import pytest
from volarb_py.factors import (
    FACTOR_COUNT,
    InvalidFactorInputsError,
    SurfacePoint,
    correlation_of,
    decompose_surface_factors,
    effective_sample_size,
    inner_product,
    lag_one_autocorrelation,
    neutralise_against_loadings,
    orthonormal_basis,
    raw_factor_vectors,
    score_residual,
)

TENORS = (0.0833, 0.25, 0.5, 1.0, 2.0)
STRIKES = (-0.4, -0.2, -0.1, 0.0, 0.1, 0.2, 0.4)
GRID = [SurfacePoint(strike, tenor) for tenor in TENORS for strike in STRIKES]
OBSERVATIONS = 200
ORTHOGONALITY_BUDGET = 1e-10
NEUTRALITY_BUDGET = 1e-12
SINGLE_EXPIRY_FACTORS = 3
DISGUISED_LEVEL_BET = 0.5
EXACT_RECONSTRUCTION = 0.999999
RESIDUAL_FLOOR = 1e-12
DISTURBED_POINT = 3
PERSISTENT_AUTOCORRELATION = 0.8
PERSISTENT_OVERSTATEMENT = 3.0
SERIES_LENGTH = 400


def factor_driven_surfaces(count: int) -> list[list[float]]:
    surfaces = []
    for index in range(count):
        phase = float(index)
        level = 0.04 * math.exp(0.2 * math.sin(phase * 0.31))
        slope = 1.0 + 0.1 * math.cos(phase * 0.17)
        skew = -0.35 + 0.08 * math.sin(phase * 0.23)
        curvature = 0.9 + 0.15 * math.cos(phase * 0.11)
        surfaces.append(
            [
                level * tenor**slope * math.exp(skew * strike + curvature * strike * strike)
                for tenor in TENORS
                for strike in STRIKES
            ]
        )
    return surfaces


def autoregressive_series(count: int, autocorrelation: float) -> list[float]:
    state = 1.0
    value = 0.0
    series = []
    for _ in range(count):
        state = math.fmod(state * 48271.0, 2147483647.0)
        shock = (state / 2147483647.0 - 0.5) * math.sqrt(12.0)
        value = autocorrelation * value + math.sqrt(1.0 - autocorrelation**2) * shock
        series.append(value)
    return series


def test_the_named_basis_is_orthonormal_on_the_grid() -> None:
    basis = orthonormal_basis(raw_factor_vectors(GRID))
    assert len(basis) == FACTOR_COUNT
    for row, one in enumerate(basis):
        for column, other in enumerate(basis):
            expected = 1.0 if row == column else 0.0
            assert inner_product(one, other) == pytest.approx(expected, abs=1e-12)


def test_a_surface_built_only_from_the_factors_leaves_no_residual() -> None:
    decomposition = decompose_surface_factors(GRID, factor_driven_surfaces(120))
    assert decomposition.identified_factor_count == FACTOR_COUNT
    assert decomposition.variance_explained > EXACT_RECONSTRUCTION
    for residual in decomposition.residuals:
        assert max(abs(value) for value in residual) < RESIDUAL_FLOOR


def test_residuals_are_orthogonal_to_the_basis_at_every_observation() -> None:
    decomposition = decompose_surface_factors(GRID, factor_driven_surfaces(60))
    basis = orthonormal_basis(raw_factor_vectors(GRID))
    for residual in decomposition.residuals:
        for vector in basis:
            assert inner_product(residual, vector) == pytest.approx(0.0, abs=ORTHOGONALITY_BUDGET)


def test_a_naive_residual_is_mostly_a_disguised_level_bet() -> None:
    surfaces = factor_driven_surfaces(OBSERVATIONS)
    decomposition = decompose_surface_factors(GRID, surfaces)
    level_series = [entry.level for entry in decomposition.loadings]

    point = 0
    average = sum(surface[point] for surface in surfaces) / len(surfaces)
    naive = [surface[point] - average for surface in surfaces]
    orthogonal = [residual[point] for residual in decomposition.residuals]

    assert abs(correlation_of(naive, level_series)) > DISGUISED_LEVEL_BET
    assert abs(correlation_of(orthogonal, level_series)) < abs(correlation_of(naive, level_series))


def test_time_domain_neutralisation_removes_what_the_cross_section_leaves() -> None:
    surfaces = factor_driven_surfaces(OBSERVATIONS)
    disturbed = [
        [
            value * (1.0 + 0.02 * math.sin(index * 0.31)) if position == DISTURBED_POINT else value
            for position, value in enumerate(surface)
        ]
        for index, surface in enumerate(surfaces)
    ]
    decomposition = decompose_surface_factors(GRID, disturbed)
    series = [residual[DISTURBED_POINT] for residual in decomposition.residuals]
    neutralised = neutralise_against_loadings(
        series, decomposition.loadings, decomposition.identified_factor_count
    )
    assert neutralised.worst_factor_correlation < NEUTRALITY_BUDGET
    assert neutralised.worst_factor_correlation < neutralised.worst_factor_correlation_before


def test_a_single_expiry_grid_cannot_identify_a_term_slope() -> None:
    grid = [SurfacePoint(strike, 0.25) for strike in STRIKES]
    observations = [
        [0.04 * math.exp(-0.3 * strike + 0.02 * index) for strike in STRIKES] for index in range(20)
    ]
    decomposition = decompose_surface_factors(grid, observations)
    assert decomposition.identified_factor_count == SINGLE_EXPIRY_FACTORS
    assert all(entry.curvature == 0.0 for entry in decomposition.loadings)


def test_autocorrelation_inflates_the_standard_error_of_a_z_score() -> None:
    for autocorrelation in (0.0, 0.5, 0.8, 0.9):
        score = score_residual(autoregressive_series(SERIES_LENGTH, autocorrelation))
        persistent = score.lag_one_autocorrelation > 0.0
        assert (score.naive_overstatement > 1.0) == persistent
        assert (score.effective_sample_size < score.observation_count) == persistent
        if persistent:
            assert abs(score.adjusted_z_score) < abs(score.naive_z_score)


def test_a_persistent_residual_is_far_less_significant_than_it_looks() -> None:
    score = score_residual(autoregressive_series(SERIES_LENGTH, 0.9))
    assert score.lag_one_autocorrelation > PERSISTENT_AUTOCORRELATION
    assert score.naive_overstatement > PERSISTENT_OVERSTATEMENT


def test_a_mean_reverting_residual_is_more_significant_than_it_looks() -> None:
    score = score_residual(autoregressive_series(SERIES_LENGTH, -0.6))
    assert score.lag_one_autocorrelation < 0.0
    assert score.effective_sample_size > score.observation_count
    assert score.naive_overstatement < 1.0


def test_the_effective_sample_size_follows_the_autocorrelation() -> None:
    assert effective_sample_size(100, 0.0) == pytest.approx(100.0)
    assert effective_sample_size(100, 0.9) == pytest.approx(100.0 * 0.1 / 1.9)
    assert lag_one_autocorrelation([3.0] * 50) == 0.0


def test_malformed_factor_inputs_are_rejected() -> None:
    with pytest.raises(InvalidFactorInputsError):
        decompose_surface_factors([SurfacePoint(0.0, 1.0)], [[1.0], [1.0]])
    with pytest.raises(InvalidFactorInputsError):
        decompose_surface_factors(GRID, [])
    with pytest.raises(InvalidFactorInputsError):
        decompose_surface_factors(GRID, [[0.04] * len(GRID), [0.04] * 3])
    with pytest.raises(InvalidFactorInputsError):
        decompose_surface_factors(GRID, [[-0.04] * len(GRID), [0.04] * len(GRID)])
    with pytest.raises(InvalidFactorInputsError):
        score_residual([1.0])
    with pytest.raises(InvalidFactorInputsError):
        decompose_surface_factors([SurfacePoint(0.0, -1.0)] * 4, [[0.04] * 4, [0.04] * 4])
